File: app/src/main/java/com/dev/moviedb/mvvm/model/movies/dto/MovieCollectionDto.java
package com.dev.moviedb.mvvm.model.movies.dto;

import com.dev.moviedb.mvvm.model.movies.Movie;
import com.dev.moviedb.mvvm.model.movies.mapper.DataMapper;

import java.util.ArrayList;
import java.util.List;

/**
 * DTO that represents the most normal results obtained via API requests.
 * Composed by a sequence of results, current page returned by the API,
 * total number of pages and total number of results.
 *
 * @version 0.0.1
 */
public class MovieCollectionDto {

    private int page;

    private List<MovieDTO> results;

    private int total_pages;

    private int total_results;

    public MovieCollectionDto(int page, List<MovieDTO> results, int totalPages, int totalResults) {
        this.page = page;
        this.results = results;
        this.total_pages = totalPages;
        this.total_results = totalResults;
    }


    public int getPage() {
        return page;
    }

    public int getTotalPages() {
        return total_pages;
    }

    public int getTotalResults() {
        return total_results;
    }

    public List<Movie> getResults() {
        List<Movie> tmpList = new ArrayList<>();
        DataMapper mapper = new DataMapper();
        for (MovieDTO dto : results){
            tmpList.add(mapper.convertFrom(dto));
        }
        return tmpList;
    }

    /**
     * To support 0.0.1 version. Used by the callbacks that talk directly to the
     * retrofit library.
     *
     */
    @Deprecated
    public List<MovieDTO> getSimpleResults() {
        return this.results;
    }
}

File: app/src/main/java/com/dev/moviedb/mvvm/model/movies/dto/MovieCollectionDTO.kt
package com.dev.moviedb.mvvm.model.movies.dto

import android.os.Parcel
import android.os.Parcelable

/**
 *
 * Yamda 1.0.0.
 */
data class MovieCollectionDTO(var page: Int,
                              var results:  List<MovieDTO>,
                              var totalPages: Int,
                              var totalResults: Int): Parcelable {
    constructor(parcel: Parcel) : this(
            parcel.readInt(),
            parcel.createTypedArrayList(MovieDTO.CREATOR),
            parcel.readInt(),
            parcel.readInt())

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        parcel.writeInt(page)
        parcel.writeTypedList(results)
        parcel.writeInt(totalPages)
        parcel.writeInt(totalResults)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<MovieCollectionDTO> {
        override fun createFromParcel(parcel: Parcel): MovieCollectionDTO {
            return MovieCollectionDTO(parcel)
        }

        override fun newArray(size: Int): Array<MovieCollectionDTO?> {
            return arrayOfNulls(size)
        }
    }
}